lib/rate-limit.js

```diff
@@ -1,86 +1,44 @@
-import { Promise } from 'es6-promise'
-
-import promisedWait from './promised-wait'
-
-/**
- * Promise-based rate limiting of a function. Assumes that the passed in
- * function will return a Promise.
- * Queues calls made to the wrapped function, and if more calls are made than
- * the allowed level of concurrency waits until a certain amount of time has
- * passed before continuing to clear calls from the queue.
- * @private
- * @param {Function} fn - call to be rate limited
- * @param {number} concurrency - Number of allowed concurrent requests
- * @param {number} delay - Delay in milliseconds for waiting after hitting the
- */
-export default function rateLimit (fn, concurrency, delay) {
-  concurrency = positiveInteger('concurrency', concurrency)
-  delay = positiveInteger('delay', delay)
-  const callQueue = []
-  let inFlight = 0
-
-  function shift () {
-    if (inFlight >= concurrency) {
-      return
+export default function rateLimit (instance, maxRetry = 10) {
+  instance.interceptors.response.use(function (response) {
+    // we don't need to do anything here
+    return response
+  }, function (error) {
+    const {config} = error
+    if (!config) {
+      return Promise.reject(error)
     }
-
-    const start = new Date().getTime()
-
-    if (callQueue.length) {
-      const call = callQueue.shift()
-      inFlight++
-
-      let result
-      try {
-        // call the wrapped method
-        const tmp = Promise.resolve(fn.apply(call.self, call.args))
-        // resolve the call with a possibly rejected promise
-        call.resolve(tmp)
-        // rate-limiting doesn't care about errors
-        result = tmp.catch(() => {})
-      } catch (err) {
-        call.reject(err)
-        result = Promise.resolve()
+    config.attemp = config.attemp || 0
+    if (error.response.status >= 500) {
+      config.attemp++
+      config.lastTimeSinceError = config.lastTimeSinceError || Date.now()
```

Comment:
Mind renaming this variable? `lastTimeSinceError` sounds like a duration already but it is instead a timestamp. Something like `lastErrorAt` or `lastErrorTimestamp` are clearer IMO